bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/handler/CaddxBridgeHandler.java

```diff
@@ -0,0 +1,408 @@
+                    partition = Integer.parseInt(caddxMessage.getPropertyById("partition_number")) + 1;
+                    break;
+                case Zone:
+                    caddxThingType = CaddxThingType.ZONE;
+                    zone = Integer.parseInt(caddxMessage.getPropertyById("zone_number")) + 1;
+                    break;
+                case Keypad:
+                    caddxThingType = CaddxThingType.KEYPAD;
+                    keypad = Integer.parseInt(caddxMessage.getPropertyById("keypad_address"));
+                    break;
+                default:
+                    logger.debug("Source has illegal value");
+                    return;
+            }
+
+            CaddxEvent event = new CaddxEvent(caddxMessage, partition, zone, keypad);
+
+            // Find the thing
+            Thing thing = findThing(caddxThingType, partition, zone, keypad);
+            CaddxDiscoveryService discoverService = getDiscoveryService();
+            if (thing != null) {
+                CaddxBaseThingHandler thingHandler = (CaddxBaseThingHandler) thing.getHandler();
+                if (thingHandler != null) {
+                    thingHandler.caddxEventReceived(event, thing);
+                }
+            } else {
+                if (discoverService != null) {
+                    discoverService.addThing(getThing(), caddxThingType, event);
+                }
+            }
+
+            // Handle specific messages that add multiple discovered things
+            if (discoverService != null) {
+                switch (caddxMessage.getCaddxMessageType()) {
+                    case Partitions_Snapshot_Message:
+                        for (int i = 1; i <= 8; i++) {
+                            if (caddxMessage.getPropertyById("partition_" + Integer.toString(i) + "_valid") == "true") {
+                                thing = findThing(CaddxThingType.PARTITION, i, null, null);
+                                if (thing != null) {
+                                    continue;
+                                }
+
+                                event = new CaddxEvent(caddxMessage, i, null, null);
+                                discoverService.addThing(getThing(), CaddxThingType.PARTITION, event);
+                            }
+                        }
+                        break;
+                    case Zones_Snapshot_Message:
+                        int zoneOffset = Integer.parseInt(caddxMessage.getPropertyById("zone_offset"));
+                        for (int i = 1; i <= 16; i++) {
+                            if (caddxMessage.getPropertyById("zone_" + Integer.toString(i) + "_trouble")
+                                    .equals("false")) {
+                                thing = findThing(CaddxThingType.ZONE, null, zoneOffset + i, null);
+                                if (thing != null) {
+                                    continue;
+                                }
```

Comment:
I wonder if it would be more performant to remove this

Reply:
See above

Reply:
Done